DOC: How to construct config files (#454)
skypyproject/skypy: docs/configuration_files.rst

Also, ``noise`` is an optional parameter and you could use its default value by omitting its definition.

  .. code:: yaml

    cosmology: !astropy.cosmology.default_cosmology.get
    z_range: !numpy.linspace [0, 2, 21]
    M_star: !astropy.modeling.models.Linear1D [-0.9, -20.4]
    phi_star: !astropy.modeling.models.Exponential1D [3e-3, -9.7]
    magnitude_limit: 23
    sky_area: 10 deg2

- `Tables and columns`:

You can create a table ``blue_galaxies`` and for now add the columns for redshift and magnitude (note here the ``schechter_lf`` returns a 2D object)

  .. code:: yaml

    tables:
      blue_galaxies:
        redshift, magnitude: !skypy.galaxies.schechter_lf
      		redshift: $z_range
      		M_star: $M_star
      		phi_star: $phi_star
      		alpha: -1.3
      		m_lim: $magnitude_limit
      		sky_area: $sky_area

`Important:` if cosmology is detected as a parameter but is not set, it automatically uses the cosmology variable defined at the top-level of the file.

This is how the entire configuration file looks like!
You may now save it as ``luminosity.yml`` and run it using the `SkyPy` `~skypy.pipeline.Pipeline`:

> .. literalinclude:: luminosity.yml

Author: Hmm not necessarily. I thought it would be clearer this way.

   :language: yaml

You may now save it as ``luminosity.yml`` and run it using the `SkyPy` `~skypy.pipeline.Pipeline`:

.. plot::
   :include-source: true
   :context: close-figs

    import matplotlib.pyplot as plt
    from skypy.pipeline import Pipeline

    # Execute SkyPy luminosity pipeline
    pipeline = Pipeline.read("luminosity.yml")
    pipeline.execute()

    # Blue population
    skypy_galaxies = pipeline['blue_galaxies']

    # Plot histograms
    fig, axs = plt.subplots(1, 2, figsize=(9, 3))

    axs[0].hist(skypy_galaxies['redshift'], bins=50, histtype='step', color='purple')
    axs[0].set_xlabel(r'$Redshift$')
    axs[0].set_ylabel(r'$\mathrm{N}$')
    axs[0].set_yscale('log')

    axs[1].hist(skypy_galaxies['magnitude'], bins=50, histtype='step', color='green')
    axs[1].set_xlabel(r'$Magnitude$')
    axs[1].set_yscale('log')

    plt.show()

> 

Reviewer: I might mention here the possibility to run the pipeline from the command line as well
```suggestion
You can also run the pipeline directly from the command line
.. code-block:: bash
    $ skypy luminosity.yml --format fits
```

Author: Goo idea

Don’t forget to check out for more complete examples_!

.. _examples: https://skypy.readthedocs.io/en/stable/examples/index.html

YAML syntax
-----------
YAML_ is a file format designed to be readable by both computers and humans.
Fundamentally, a file written in YAML consists of a set of key-value pairs.
Each pair is written as ``key: value``, where whitespace after the ``:`` is optional.

This guide introduces the main syntax of YAML relevant when writing
a configuration file to use with ``SkyPy``. Essentially, it begins with
definitions of individual variables at the top level, followed by the tables,
and, within the table entries, the features of objects to simulate are included.
Main keywords: ``parameters``, ``cosmology``, ``tables``.

Variables
^^^^^^^^^

> * `Define a variable`: a variable is defined as a key-value pair at the top of the file. You can define any type of variable:

Reviewer 2: This is not quite accurate, you cannot define "any type" of variable without calling a constructor. pyyaml will interpret any numeric data with the appropriate type (int, float, bool). Similarly for lists and dicts. We add a definition where any string that astropy can interpret as a Quantity will be converted to and stored as a Quantity (you should add this to your example). Everything else is stored as a string (wether or not you explicitly use quotes).

  .. code:: yaml

      counter: 100 # An integer
      miles: 1000.0 # A floating point
      name: "Joy" # A string
      mylist: [ 'abc', 789, 2.0e3 ] # A list

> * `Reference a variable`: variables can be referenced by their full name tagged with a dollar sign ``$``.

Reviewer 2: I would suggest moving the discussion of variable referencing to the "Functions" section. The example you give of creating one variable that is a reference to another doesn't really achieve anything meaningful and doesn't motivate the user as to why they would want to use this feature. In the functions section however you can give a concrete example of passing one variable as an argument to another function which is exactly how we intend referencing to be used.

  For example, if you previously defined a list of galaxy properties:

  .. code:: yaml

      galaxy_properties: [ 'OBJECT_ID', 'RA', 'DEC', 'REDSHIFT', 'FLUX', 'FLUX_ERR' ]

  You could reference the variable:

  .. code:: yaml

      catalog: $galaxy_properties

Functions
^^^^^^^^^
* `Call a function`: functions are defined as tuples where the first entry is the fully qualified function name tagged with and exclamation mark ``!`` and the second entry is either a list of positional arguments or a dictionary of keyword arguments.

  For example, if you need to call the ``log10()`` and ``linspace()`` NumPy_ functions, then you define the following key-value pairs:

  .. code:: yaml

      log_of_2: !numpy.log10 [2]
      myarray: !numpy.linspace [0, 2.5, 10]

  You can also define parameters of functions with a dictionary of keyword arguments.
  Imagine you want to compute the comoving distance for a range of redshifts and an `Astropy` Planck 2015 cosmology.
  To run it with the `SkyPy` pipeline, call the function and define the parameters as an indented dictionary.

  .. code:: yaml

      comoving_distance: !astropy.cosmology.Planck15.comoving_distance

>         z: !numpy.linspace [ 0, 1.3, num=10 ]

Reviewer 2: This example doesn't work as intended. You can't give both args and kwargs in our syntax. Either you give a list of args, or a dict of kwargs. Running this I get an error `TypeError: 'str' object cannot be interpreted as an integer` because it is trying to pass the string "num=10" as the third arg, but the function expects an integer. We should make this point more clearly in the text. The correct way to run this function as you intend is either:

```
      comoving_distance: !astropy.cosmology.Planck15.comoving_distance
        z: !numpy.linspace [0, 1.3, 10]
```

or

```
      comoving_distance: !astropy.cosmology.Planck15.comoving_distance
        z: !numpy.linspace {start: 0, stop: 1.3, num: 10}
```

  referencing the job that initialises the empty table ``halos`` and ``galaxies``.
  The potential risk is that the function could be called before the jobs ``halos`` and ``galaxies`` are finished, so the tables would be empty

  To overcome this issue you can initialise your ``matching`` table with ``init``, specify their dependences with the keyword ``depends``
  and ensure the tables are completed before calling the function with ``.complete``. The previous configuration file reads now:

  .. code:: yaml

    tables:
      halos:
        halo_mass: !mylibrary.my_halo_mass_function
          m_min: 1.0e8
          m_max: 1.0e14
      galaxies:
        luminosity: !mylibrary.my_schechter_function
          alpha: 1.20
      matching:
        init: !numpy.vstack
          tup: [ $halos, $galaxies ]
          depends: [ tup.complete ]

Cosmology, a special parameter
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

* `Define parameters`: parameters are variables that can be modified at execution.

  For example,

  .. code:: yaml

    parameters:
      hubble_constant: 70
      omega_matter: 0.3

> * The `cosmology` to be used by functions within the pipeline only needs to be set up at the top. If a function needs ``cosmology`` as an input, you need not define it again, it is automatically detected.

Author: I encountered the same problem with Planck18, but I think you can do any other cosmology. I think it is an issue related to the Astropy version(?)

> * The `cosmology` to be used by functions within the pipeline only needs to be set up at the top. If a function needs ``cosmology`` as an input, you need not define it again, it is automatically detected.

Author: `Planck18.clone`

> * The `cosmology` to be used by functions within the pipeline only needs to be set up at the top. If a function needs ``cosmology`` as an input, you need not define it again, it is automatically detected.

Reviewer 2: #463 will allow you to directly import cosmologies without any hack:
```yaml
cosmology: !astropy.cosmology.Planck18
```

  but the mean depends on how massive the galaxies are:

  .. code:: yaml

    tables:
      galaxies:
        mass: !numpy.random.lognormal
          mean: 5.
          size: 10000
        radius: !numpy.random.lognormal
          mean: $galaxies.mass

* `Multi-column assignment`: multi-column assignment is performed with any 2d-array, where one of the dimensions is interpreted
  as the rows of the table and the second dimension, as separate columns. Or you can do it from a function that returns a tuple.

  We use multi-column assignment in the following example where we sample different halo properties from a lognormal distribution (returns a tuple of three arrays):

  .. code:: yaml

    tables:
      halos:
        mass, radius, concentration: !numpy.random.lognormal
          size: [10000, 3]

* `Table initialisation`: by default tables are initialised using ``astropy.table.Table()`` however this can be overridden using the ``.init`` keyword to initialise the table with any function call.

  For example, you can stack galaxy properties such as radii and mass:

  .. code:: yaml

    radii: !numpy.logspace [ 1, 2, 100 ]
    mass: !numpy.logspace [ 9, 12, 100 ]

>     tables:
>       .init: !astropy.table.vstack
>         galaxies: [ $radii, $mass ]

Reviewer 2: ```suggestion
tables:
  galaxies:
    .init: !astropy.table.hstack [[ $radii, $mass ]]
```

* `Table reference`: when a function call depends on tables, you need to ensure the referenced table has the necessary content and is not empty.
  You can do that with ``.complete``.

  Example: you want to perform a very simple abundance matching, i.e. painting galaxies within your halos.
  You can create two tables ``halos`` and ``galaxies`` storing the halo mass and galaxy luminosities.
  Then you can stack these two tables and store it in a third table called ``matching``.

  .. code:: yaml

    tables:
      halos:
        halo_mass: !numpy.random.uniform
          low: 1.0e8
          high: 1.0e14
          size: 20
      galaxies:
        luminosity: !numpy.random.uniform
          low: 0.05
          high: 10.0
          size: 20
      matching:

>         .init: !astropy.table.vstack

Reviewer 2: ```suggestion
        .init: !astropy.table.hstack
```

Reviewer 2: However this example still will not work until #464 is resolved

In our example, we are using ``Astropy`` linear and exponential models for the characteristic absolute magnitude and the amplitude, respectively.
Also, ``noise`` is an optional parameter and you could use its default value by omitting its definition.

  .. code:: yaml

    cosmology: !astropy.cosmology.default_cosmology.get
    z_range: !numpy.linspace [0, 2, 21]
    M_star: !astropy.modeling.models.Linear1D [-0.9, -20.4]
    phi_star: !astropy.modeling.models.Exponential1D [3e-3, -9.7]
    magnitude_limit: 23
    sky_area: 10 deg2

- `Tables and columns`:

You can create a table ``blue_galaxies`` and for now add the columns for redshift and magnitude (note here the ``schechter_lf`` returns a 2D object)

  .. code:: yaml

    tables:
      blue_galaxies:
        redshift, magnitude: !skypy.galaxies.schechter_lf
      		redshift: $z_range
      		M_star: $M_star
      		phi_star: $phi_star
      		alpha: -1.3
      		m_lim: $magnitude_limit
      		sky_area: $sky_area

`Important:` if cosmology is detected as a parameter but is not set, it automatically uses the cosmology variable defined at the top-level of the file.

This is how the entire configuration file looks like!

> You may now save it as ``luminosity.yml`` and run it using the `SkyPy` `~skypy.pipeline.Pipeline`:

Reviewer 3: ```suggestion
```
See other comment.

.. literalinclude:: luminosity.yml
   :language: yaml

> You may now save it as ``luminosity.yml`` and run it using the `SkyPy` `~skypy.pipeline.Pipeline`:

Reviewer 3: This sentence is twice. Also in line 59. Is this on purpose?
I think we want to keep this line and remove 59?

.. code-block:: bash

    $ skypy luminosity.yml --format fits

Don’t forget to check out for more complete examples_!

.. _examples: https://skypy.readthedocs.io/en/stable/examples/index.html

YAML syntax
-----------
YAML_ is a file format designed to be readable by both computers and humans.
Fundamentally, a file written in YAML consists of a set of key-value pairs.
Each pair is written as ``key: value``, where whitespace after the ``:`` is optional.

This guide introduces the main syntax of YAML relevant when writing
a configuration file to use with ``SkyPy``. Essentially, it begins with
definitions of individual variables at the top level, followed by the tables,
and, within the table entries, the features of objects to simulate are included.
Main keywords: ``parameters``, ``cosmology``, ``tables``.

Variables
^^^^^^^^^
* `Define a variable`: a variable is defined as a key-value pair at the top of the file.
  YAML is able to interpret any numeric data with the appropriate type: integer, float, boolean.
  Similarly for lists and dictionaries.
  In addition, SkyPy has added extra functionality to interpret and store Astropy Quantities_.

>   Everything else is stored as a string (whit or without explicitly using quotes)

Reviewer 3: ```suggestion
  Everything else is stored as a string (with or without explicitly using quotes)
```

    Imagine you want to add a column for the telescope collecting surface

  .. code:: yaml

      tables:
        telescope:
          spectral_lines: !scipy.stats.norm.rvs
            loc: 550
            scale: 1.6
            size: 100
          collecting_surface: !numpy.random.uniform
            low: 6.9
            high: 7.1
            size: 100

* `Reference a column`: columns in the pipeline can be referenced by their full name tagged with a dollar sign ``$``.
  Example: the radius of cosmic voids seem to follow a lognormal distribution. You can create a table ``cosmic_voids``
  with a column ``radii`` where you sample 10000 void sizes and two other columns, ``mean`` and ``variance`` that reference
  the first column

  .. code:: yaml

    tables:
      cosmic_voids:
        radii: !scipy.stats.lognorm.rvs
          s: 5.
          size: 10000
        mean: !numpy.mean
          a: $cosmic_voids.radii
        variance: !numpy.var
          a: $cosmic_voids.radii

* `Multi-column assignment`: if a function returns multiple columns, you can chose to assign them to multiple columns with different names or to a muti-column object.

> 

Author: rephrase

  Imagine you want the distribution of the circular velocities of 1000 halos following a Maxwellian distribution.
  The histogram NumPy_ returns a 2-dimensional object. Therefore, you can choose

  .. code:: yaml

    tables:
      halos:
        circular_velocities: !scipy.stats.maxwell.rvs
          s: 250
          size: 1000
        hist, bin_edges: !numpy.histogram
          a: $halos.circular_velocities
          density: True

  or a multi-column assignment

  .. code:: yaml

    tables:
      halos:
        circular_velocities: !scipy.stats.maxwell.rvs
          s: 250
          size: 1000
        histogram: !numpy.histogram
          a: $halos.circular_velocities
          density: True

> 

Author: not multicolumn-assignment. The example above is the multicolumn assignment.

    pipeline = Pipeline.read("luminosity.yml")
    pipeline.execute()

    # Blue population
    skypy_galaxies = pipeline['blue_galaxies']

    # Plot histograms
    fig, axs = plt.subplots(1, 2, figsize=(9, 3))

    axs[0].hist(skypy_galaxies['redshift'], bins=50, histtype='step', color='purple')
    axs[0].set_xlabel(r'$Redshift$')
    axs[0].set_ylabel(r'$\mathrm{N}$')
    axs[0].set_yscale('log')

    axs[1].hist(skypy_galaxies['magnitude'], bins=50, histtype='step', color='green')
    axs[1].set_xlabel(r'$Magnitude$')
    axs[1].set_yscale('log')

    plt.tight_layout()
    plt.show()

You can also run the pipeline directly from the command line and write the outputs to a fits file:

.. code-block:: bash

    $ skypy luminosity.yml luminosity.fits

Don’t forget to check out for more complete examples_!

.. _examples: https://skypy.readthedocs.io/en/stable/examples/index.html

YAML syntax
-----------
YAML_ is a file format designed to be readable by both computers and humans.
Fundamentally, a file written in YAML consists of a set of key-value pairs.
Each pair is written as ``key: value``, where whitespace after the ``:`` is optional.

> 

Reviewer 2: ```suggestion
The hash character (#) denotes the start of a comment and all further text on that
line is ignored by the parser.

```

Main keywords: ``parameters``, ``cosmology``, ``tables``.

Variables
^^^^^^^^^
* `Variable definition`: a variable is defined as a key-value pair at the top of the file.
  YAML is able to interpret any numeric data with the appropriate type: integer, float, boolean.
  Similarly for lists and dictionaries.
  In addition, SkyPy has added extra functionality to interpret and store Astropy Quantities_.
  Everything else is stored as a string (with or without explicitly using quotes)

  .. code:: yaml

      # YAML interprets
      counter: 100 # An integer
      miles: 1000.0 # A floating point
      name: "Joy" # A string
      planet: Earth # Another string
      mylist: [ 'abc', 789, 2.0e3 ] # A list
      mydict: { 'fruit': 'orange', 'year': 2020 } # A dictionary

      # SkyPy extra functionality
      angle: 10 deg
      distance: 300 kpc

* `Import objects`:
  the SkyPy configuration syntax allows objects to be imported directly from external
  (sub)modules using the ``!`` tag and providing neither a list of arguments or a
  dictionary of keywords. For example, this enables the import and usage of any Astropy cosmology:

  .. code:: yaml

>       cosmo1: !astropy.cosmology.Planck13 # import the Planck13 object and bind to cosmo1
>       cosmo2: !astropy.cosmology.default_cosmology.get [] # call default_cosmology.get() and bind to cosmo2

Reviewer 2: ```suggestion
      cosmology: !astropy.cosmology.Planck13 # import the Planck13 object and bind it to the variable named "cosmology"
```

* `Variable reference`: variables can be referenced by their full name tagged with a dollar sign ``$``.
  In the previous example you could also define the variables at the top-level of the file and then reference them:

  .. code:: yaml

      redshift: !numpy.linspace [ 0, 1.3, 10 ]
      comoving_distance: !astropy.cosmology.Planck15.comoving_distance
        z: $redshift

* The `cosmology` to be used by functions within the pipeline only needs to be set up at the top. If a function needs ``cosmology`` as an input, you need not define it again, it is automatically detected.

  For example, calculate the angular size of a galaxy with a given physical size, at a fixed redshift and for a given cosmology:

  .. code:: yaml

      cosmology: !astropy.cosmology.FlatLambdaCDM
        H0: 70
        Om0: 0.3
      size: !skypy.galaxies.morphology.angular_size
        physical_size: 10 kpc
        redshift: 0.2

* `Job completion`: ``.depends`` can be used to force any function call to wait for completion
  of any other job.

  A simple example where, for some reason, the comoving distance needs to be called after
  completion of the angular size function:

  .. code:: yaml

>     size: !skypy.galaxies.morphology.angular_size

Reviewer 2: ```suggestion
    cosmology: !astropy.cosmology.Planck15
    size: !skypy.galaxies.morphology.angular_size
```

  .. code:: yaml

      tables:
        telescope:
          spectral_lines: !scipy.stats.norm.rvs
            loc: 550
            scale: 1.6
            size: 100
          collecting_surface: !numpy.random.uniform
            low: 6.9
            high: 7.1
            size: 100

* `Column reference`: columns in the pipeline can be referenced by their full name tagged with a dollar sign ``$``.
  Example: the galaxy mass that follows a lognormal distribution. You can create a table ``galaxies``
  with a column ``mass`` where you sample 10000 object and a second column, ``radius`` which also follows a lognormal distribution
  but the mean depends on how massive the galaxies are:

  .. code:: yaml

    tables:
      galaxies:
        mass: !numpy.random.lognormal
          mean: 5.
          size: 10000
        radius: !numpy.random.lognormal
          mean: $galaxies.mass

* `Multi-column assignment`: multi-column assignment is performed with any 2d-array, where one of the dimensions is interpreted
  as the rows of the table and the second dimension, as separate columns. Or you can do it from a function that returns a tuple.

>   We use multi-column assignment in the following example where we sample different halo properties from a lognormal distribution (returns a tuple of three arrays):

Reviewer 2: ```suggestion
  We use multi-column assignment in the following example where we sample a two-dimensional array of values from a lognormal distribution and then store them as three columns in a table:
```

###################

> Configuration files

Reviewer 2: ```suggestion
Configuration Files
```

In this example, we sample redshifts and magnitudes from the SkyPy luminosity function, `~skypy.galaxies.schechter_lf`.

- `Define variables`:

From the documentation, the parameters for the `~skypy.galaxies.schechter_lf` function are: ``redshift``, the characteristic absolute magnitude ``M_star``, the amplitude ``phi_star``, faint-end slope parameter ``alpha``,
the magnitude limit ``magnitude_limit``, the fraction of sky ``sky_area``, ``cosmology`` and ``noise``.
If you are planning to reuse some of these parameters, you can define them at the top-level of your configuration file.
In our example, we are using ``Astropy`` linear and exponential models for the characteristic absolute magnitude and the amplitude, respectively.
Also, ``noise`` is an optional parameter and you could use its default value by omitting its definition.

  .. code:: yaml

    cosmology: !astropy.cosmology.default_cosmology.get
    z_range: !numpy.linspace [0, 2, 21]
    M_star: !astropy.modeling.models.Linear1D [-0.9, -20.4]
    phi_star: !astropy.modeling.models.Exponential1D [3e-3, -9.7]
    magnitude_limit: 23

>     sky_area: 10 deg2

Reviewer 2: ```suggestion
    sky_area: 0.1 deg2
```
I suggest a smaller sky area to reduce the run time and the size of the output file for anybody running this example

  Imagine you want to compute the comoving distance for a range of redshifts and an `Astropy` Planck 2015 cosmology.
  To run it with the `SkyPy` pipeline, call the function and define the parameters as an indented dictionary.

  .. code:: yaml

      comoving_distance: !astropy.cosmology.Planck15.comoving_distance
        z: !numpy.linspace [ 0, 1.3, 10 ]

  Similarly, you can specify the functions arguments as a dictionary:

  .. code:: yaml

      comoving_distance: !astropy.cosmology.Planck15.comoving_distance
        z: !numpy.linspace {start: 0, stop: 1.3, num: 10}

  `N.B.` To call a function with no arguments, you should pass an empty list of
  ``args`` or an empty dictionary of ``kwargs``. For example:

  .. code:: yaml

      cosmo: !astropy.cosmology.default_cosmology.get []

* `Reference a variable`: variables can be referenced by their full name tagged with a dollar sign ``$``.
  In the previous example you could also define the variables at the top-level of the file and then reference them:

  .. code:: yaml

      redshift: !numpy.linspace [ 0, 1.3, 10 ]
      comoving_distance: !astropy.cosmology.Planck15.comoving_distance
        z: $redshift

>   Please, check below in the `cosmology`_ section how to use different cosmologies with the SkyPy pipeline.

Reviewer 2: Instead of a link pointing somewhere else in the documentation, it would make sense to move that information here under a new "cosmology" bullet point

Tables
^^^^^^

* `Create a table`: a dictionary of table names, each resolving to a dictionary of column names for that table.

  Let us create a table called ``telescope`` with a column to store the width of spectral lines that follow a normal distribution

  .. code:: yaml

      tables:
        telescope:

>           spectral_lines: !scipy.stats.norm

Reviewer 2: ```suggestion
          spectral_lines: !scipy.stats.norm.rvs
```

            loc: 550
            scale: 1.6
            size: 100

* `Add a column`: you can add as many columns to a table as you need.
    Imagine you want to add a column for the telescope collecting surface

  .. code:: yaml

      tables:
        telescope:

>           spectral_lines: !scipy.stats.norm

Reviewer 2: ```suggestion
          spectral_lines: !scipy.stats.norm.rvs
```

            loc: 550
            scale: 1.6
            size: 100
          collecting_surface: !numpy.random.uniform
            low: 6.9

>             high: 7.1

Reviewer 2: ```suggestion
            high: 7.1
            size: 100
```
This example doesn't work because the columns don't have the same length; is this what you intended?

* `Reference a column`: columns in the pipeline can be referenced by their full name tagged with a dollar sign ``$``.
  Example: the radius of cosmic voids seem to follow a lognormal distribution. You can create a table ``cosmic_voids``
  with a column ``radii`` where you sample 10000 void sizes and two other columns, ``mean`` and ``variance`` that reference
  the first column

  .. code:: yaml

    tables:
      cosmic_voids:
        radii: !scipy.stats.lognorm.rvs

>           s: 200.

Reviewer 2: ```suggestion
          s: 5.
```
`s=200.` is too large and leads to `RuntimeWarning: overflow encountered in exp`

          size: 10000
        mean: !numpy.mean
          a: $cosmic_voids.radii
        variance: !numpy.var

>           a: $cosmic_voids.radii

Reviewer 2: `mean` and `variance` are summary statistics each returning a single number and don't make sense as columns? Currently cosmic_voids.mean is just the mean value repeated 10000 times, and similarly for variance.

Author: True. Let me think about this

* `Multi-column assignment`: if a function returns multiple columns, you can chose to assign them to multiple columns with different names or to a muti-column object.

  Imagine you want the distribution of the circular velocities of 1000 halos following a Maxwellian distribution.
  The histogram NumPy_ returns a 2-dimensional object. Therefore, you can choose

  .. code:: yaml

    tables:
      halos:
        circular_velocities: !scipy.stats.maxwell.rvs
          s: 250
          size: 1000
        hist, bin_edges: !numpy.histogram
          a: $halos.circular_velocities

>           density: True

Reviewer 2: - `s` is not a valid keyword argument for `scipy.stats.maxwell.rvs`
- `hist` and `bin_edges have different lengths from each other and from `circular_velocities` and so cannot be valid columns in the same table
- Conceptually this example makes no sense. It is supposed to be a table where each row represents a dark matter halo and each column is a property of each halo e.g. mass, circular velocity, velocity dispersion, number of subhalos etc. A histogram (and the bin edges) is not a halo property!

  or a multi-column assignment

  .. code:: yaml

    tables:
      halos:
        circular_velocities: !scipy.stats.maxwell.rvs
          s: 250
          size: 1000
        histogram: !numpy.histogram
          a: $halos.circular_velocities
          density: True

* `Referencing tables: table.init and table.complete dependencies`.

  There are times when your function depends on tables. In these
  cases, you need to ensure the referenced table has the necessary content and is not empty.

  Example: you want to perform a very simple abundance matching, i.e. painting galaxies within your halos.
  You can create two tables ``halos`` and ``galaxies`` storing the halo mass and galaxy luminosities.
  Then you can stack these two tables and store it in a third table called ``matching``.

  `Beware`: referencing tables is a bit different to variables or columns
  -- where you simply tag a dollar sign.

  The challenge: let ``tableA`` be a table with columns ``c1`` and ``c2``.
  In configuration language, ``tableA`` is the name of the job.

>   *That means, when executing the configuration file, the first thing that happens is call ``tableA``, second,  call ``tableA.c1`` and third, call ``tableA.c2``.*

Reviewer 2: Technically not true: if there is no dependency between `tableA.c1` and `tableA.c2` then we cannot say which one will run first

  If you used the dollar sign to reference your ``tableA`` inside a function, this function might be called before the job ``tableA`` is complete, and the table will be empty.

>   `The solution`: to correctly reference tables, initialise your table with ``init``, specify their dependences with the keyword ``depends``

Reviewer 2: ```suggestion
  `The solution`: to correctly reference tables, initialise your table with ``.init``, specify their dependences with the keyword ``.depends``
```

>   and ensure the tables are completed before calling the function with ``.complete``. Our example:

Reviewer 2: This example if overly complicated; it is introducing multiple new features and conflating them when they are in fact separate:
- `.depends` can be used to force any function call to wait on the completion of any other job (probably should be mentioned as a bullet point in the "functions" section)
- Every table has an implicit `table_name.complete` job that occurs after all of its columns have been generated but does nothing
- By default tables are initialised using `astropy.table.Table()` however this can be overridden using the `.init` keyword to initialise the table with any function call

  .. code:: yaml

    tables:
      halos:
        halo_mass: !numpy.random.uniform
          low: 1.0e8
          high: 1.0e14
          size: 20
      galaxies:
        luminosity: !numpy.random.uniform
          low: 0.05
          high: 10.0
          size: 20
      matching:

>         init: !numpy.vstack
>           tuple: [ $halos, $galaxies ]
>           depends: [ tuple.complete ]

Reviewer 2: ```suggestion
        .init: !astropy.table.vstack
          tables: [ $halos, $galaxies ]
          .depends: [ halos.complete, galaxies.complete ]
```

> Import objects

Reviewer 2: I think how to import objects would make more sense as second bullet point under "Variables"?

^^^^^^^^^^^^^^
* The SkyPy configuration syntax allows objects to be imported directly from external
  (sub)modules using the ``!`` tag and providing neither a list of arguments or a
  dictionary of keywords. For example, this enables the import and usage of any Astropy cosmology:

  .. code:: yaml

    cosmo1: !astropy.cosmology.Planck13 # import the Planck13 object and bind to cosmo1
    cosmo2: !astropy.cosmology.default_cosmology.get [] # call default_cosmology.get() and bind to cosmo2

Cosmology
^^^^^^^^^

* `Define parameters`: parameters are variables that can be modified at execution.

  For example,

  .. code:: yaml

    parameters:
      hubble_constant: 70

>       omega_matter: 0.3

Reviewer 2: "Parameters" should probably be it's own section (it has nothing explicitly to do with "cosmology") with at least a brief explanation of what is different between a parameter and a normal variable

* The `cosmology` to be used by functions within the pipeline only needs to be set up at the top. If a function needs ``cosmology`` as an input, you need not define it again, it is automatically detected.

  .. code:: yaml

    parameters:
      hubble_constant: 70
      omega_matter: 0.3
    cosmology: !astropy.cosmology.FlatLambdaCDM
      H0: $hubble_constant

>       Om0: $omega_matter

Reviewer 2: This example does not demonstrate what is discussed in the text

  If you used the dollar sign to reference your ``tableA`` inside a function, this function might be called before the job ``tableA`` is complete, and the table will be empty.

  `The solution`: to correctly reference tables, initialise your table with ``init``, specify their dependences with the keyword ``depends``
  and ensure the tables are completed before calling the function with ``.complete``. Our example:

  .. code:: yaml

    tables:
      halos:
        halo_mass: !numpy.random.uniform
          low: 1.0e8
          high: 1.0e14
          size: 20
      galaxies:
        luminosity: !numpy.random.uniform
          low: 0.05
          high: 10.0
          size: 20
      matching:
        init: !numpy.vstack
          tuple: [ $halos, $galaxies ]
          depends: [ tuple.complete ]

  `A non-working example`: this is an example of an incorrect table referencing.

  .. code:: yaml

    tables:
    halos:
      halo_mass: !numpy.random.uniform
        low: 1.0e8
        high: 1.0e14
        size: 20
    galaxies:
      luminosity: !numpy.random.uniform
        low: 0.05
        high: 10.0
        size: 20
      matching_wrong:
        match: !numpy.vstack

>           tuple: [ $halos, $galaxies ]

Reviewer 2: I generally don't like showing examples of how not to do something and would suggest removing this

  Explanation: when calling the function ``numpy.vstack()`` and referencing the tables ``$halos`` and ``$galaxies``, this is actually
  referencing the job that initialises the empty table ``halos`` and ``galaxies``.
  The ``numpy.vstack()`` function is called before the jobs ``halos`` and ``galaxies`` are finished, so the tables are empty.

Import objects
^^^^^^^^^^^^^^
* The SkyPy configuration syntax allows objects to be imported directly from external
  (sub)modules using the ``!`` tag and providing neither a list of arguments or a
  dictionary of keywords. For example, this enables the import and usage of any Astropy cosmology:

  .. code:: yaml

>     cosmo1: !astropy.cosmology.Planck13 # import the Planck18 object and bind to cosmo1

Reviewer 2: ```suggestion
    cosmo1: !astropy.cosmology.Planck13 # import the Planck13 object and bind to cosmo1
```
